### ISSUE TITLE
bloscpack locked in multiprocessing mode

### ISSUE THREAD

**Issue by erogol** [Blosc/bloscpack#58]
Here I share a mimicry of my problem where the code freezes in multiprocess mode and works just fine with single process. Below code tries to create a dummy array simulating my image files and tries to save them with multi processing. If you reduce number of processes one, it just works fine but with >1 processes it freezes. 

I use Ubuntu 16.04 , Python 3.6 with Conda. 

```

import os
import sys
import tempfile

import numpy as np
import bloscpack as bp

from tqdm import tqdm
from concurrent.futures import ProcessPoolExecutor, as_completed


def parallel_process(array, function, n_jobs=16, use_kwargs=False, front_num=3):
    """
        A parallel version of the map function with a progress bar. 

        Args:
            array (array-like): An array to iterate over.
            function (function): A python function to apply to the elements of array
            n_jobs (int, default=16): The number of cores to use
            use_kwargs (boolean, default=False): Whether to consider the elements of array as dictionaries of 
                keyword arguments to function 
            front_num (int, default=3): The number of iterations to run serially before kicking off the parallel job. 
                Useful for catching bugs
        Returns:
            [function(array[0]), function(array[1]), ...]
    """
    #We run the first few iterations serially to catch bugs
    if front_num > 0:
        front = [function(**a) if use_kwargs else function(a) for a in array[:front_num]]
    #If we set n_jobs to 1, just run a list comprehension. This is useful for benchmarking and debugging.
    if n_jobs==1:
        return front + [function(**a) if use_kwargs else function(a) for a in tqdm(array[front_num:])]
    #Assemble the workers
    with ProcessPoolExecutor(max_workers=n_jobs) as pool:
        #Pass the elements of array into function
        if use_kwargs:
            futures = [pool.submit(function, **a) for a in array[front_num:]]
        else:
            futures = [pool.submit(function, a) for a in array[front_num:]]
        kwargs = {
            'total': len(futures),
            'unit': 'it',
            'unit_scale': True,
            'leave': True
        }
        #Print out the progress as tasks complete
        for f in tqdm(as_completed(futures), **kwargs):
            pass
    out = []
    #Get the results from the futures. 
    for i, future in tqdm(enumerate(futures)):
        try:
            out.append(future.result())
        except Exception as e:
            out.append(e)
    return front + out


def dump_blosc(data, filename):
    with open(filename, 'wb') as f:
        f.write(bp.pack_ndarray_str(data))


def write_data(inputs):
    dummy = np.random.rand(16,3,224,224).astype('uint8')
    tf = tempfile.NamedTemporaryFile()
    dump_blosc(dummy, tf.name)


if __name__ == '__main__':
    parallel_process(range(100), write_data, n_jobs=2)
    # for dir in dirs:
    #     print(dir)
    #     write_data(dir)

```

**Comment by esc** [comment]
Using the following alternative works fine, using the older `multiprocessing` package.

```
    from multiprocessing.pool import Pool
    p = Pool(4)
    p.map(write_data, tqdm(range(100000000)))
```

I recently also had a very confusing encounter with `concurrent.futures` --- I'll have to understand the library in more detail to be able to debug the above code.

**Comment by esc** [comment]
Also, for the working case where `n_jobs=1` the code for `parallel_process` bypasses the `concurrent.futures` stuff completely, which would explain why it works.

**Comment by esc** [comment]
Also, the following code seems to work:

```
    with ProcessPoolExecutor(max_workers=4) as e:
        res = list(e.map(write_data, tqdm(range(100000000))))
```

**Comment by esc** [comment]
Should we continue to investigate this?

**Comment by ogrisel** [comment]
I am not sure yet what the underlying code of the problem is, but it works with loky:
```
pip install loky
```

then replace in your code:

```python
from loky import get_reusable_executor

...
pool = get_reusable_executor(max_workers=n_jobs)
```

You can drop `the` with statement as loky's executor instance is meant to be reused over and over again: https://github.com/tomMoral/loky


ping @tomMoral: we should have a look at this case on Friday.

**Comment by tomMoral** [comment]
Your freeze is due to bad interaction between `fork` and `bloscpack`. Actualy, your code works if you use spawned processes such as:

```python
import multiprocessing

...

if __name__ == "__main__":
    multiprocessing.set_start_method("spawn")
    ...
```

Also, it does not freeze with `front_num=0` and `fork`, meaning that you have problems similar to the one with `openMP`.  
`loky` solves that by using `spawn` processes by default.

**Comment by esc** [comment]
Interesting, `bloscpack` uses `python-blosc` under the hood and this releases the GIL (IIRC), could that have something to do with it?

**Comment by esc** [comment]
cc @guillaumebrg this might be interesting for you too, in case you ever want to run bloscpack in parallel.

**Comment by FrancescAlted** [comment]
@esc By default `python-blosc` runs with the GIL.  If you want to release the GIL, then use [blosc.set_releasegil()](http://python-blosc.blosc.org/reference.html#blosc.set_releasegil).

**Comment by tomMoral** [comment]
I do not think it is related to the GIL release as your code is running on separated processes that do not share the GIL.
My 2c on the issue: under the hood, `python-blosc` seems to rely on threads that are re-used for multiple computations. When you use `front_num > 0`, the threads are started in the main program. When you call `fork`, the library behaves like if it had already started its worker threads in the subprocess even though these threads do not exist, because of the behavior of `fork` which breaks POSIX conventions. This result in a freeze, as your subprocesses wait for results from phantom threads.

**Comment by FrancescAlted** [comment]
@tomMoral Yes, you are right in `python-blosc` (`c-blosc` actually) relying on threads that are re-used for multiple computations.  This is where `blosc.set_releasegil(True)` may help: this releases the GIL, but at the same time calls a specific version of the Blosc compression routine, alas [blosc_decompress_ctx()](https://github.com/Blosc/python-blosc/commit/9fb4377ce46b74ea6f1aab4d7e276dce693d158a#diff-30554da7c9bdba4fdb78c7aa92dab893L366), that activates an internal lock (beware, not the GIL) so as to allow a smooth operation when called from a multithreaded app.

Message to take home: for multi-threaded apps, call `blosc.set_releasegil(True)` prior than anything.

**Comment by ogrisel** [comment]
Note that the above code does not use blosc at all. To me it's a potential bug in `ProcessPoolExecutor` itself possibly caused by some unexpected interaction with tqdm.

**Comment by esc** [comment]
Yeah, we usually want to wrap this stuff in a progress bar.

@erogol maybe you could try to release the GIL in bloscpack as @FrancescAlted suggested.

**Comment by ogrisel** [comment]
Also note that it's weird to access the result of `Future` after the shutdown of the executor (that is: outside of the `with` block). I tried to move the result fetching back into the with block but that did not fix the freeze. There is another issue. Anyway the executor should raise an catchable exception instead of freezing the Python interpreter in case of bad usage.

**Comment by esc** [comment]
FWIW: I have had problems in a different code-base too, with exceptions being raised in the workers, might try `loky` there too.

**Comment by ogrisel** [comment]
Indeed I thought that `write_data(dir)` was commented out. I did not read correctly. Ok so @tomMoral's analysis is probably correct. The issue is that `ProcessPoolExecutor` uses the `fork` start method which is not POSIX compliant. loky fixes that and the fix will be contributed upstream. Releasing the GIL in blosc will not fix the issue.

It would be possible to make blosc more robust to this family of issues by shutting down its internal thread pool before any fork using the `pthread_atfork` hook.

**Comment by ogrisel** [comment]
@FrancescAlted @esc If you want to fix the issue in blosc you can take inspiration from https://github.com/xianyi/OpenBLAS/pull/343

**Comment by FrancescAlted** [comment]
@ogrisel Thanks for the suggestion on your OpenBLAS fix.  However, I'd prefer to wait until someone confirms that `blosc.set_releasegil(True)` is not helping.

**Comment by ogrisel** [comment]
This is unrelated. Releasing the GIL is only useful to make Python level thread based parallelism more efficient.

**Comment by ogrisel** [comment]
Actually @FrancescAlted you are right, releasing the GIL in blosc fixes the issue as well. I am not sure why this is the case.

**Comment by FrancescAlted** [comment]
Great!  You probably read too fast over this:

This is where blosc.set_releasegil(True) may help: this releases the GIL, **but at the same time calls a specific version of the Blosc compression routine, alas blosc_decompress_ctx(), that activates an internal lock (beware, not the GIL) so as to allow a smooth operation when called from a multithreaded app**.

At least it makes sense to me ;) .

**Comment by ogrisel** [comment]
The problem with using the multiprocessing "fork" start method as is the case in all current versions of `concurrent.futures` is that any pthread pool initialized in a C runtime library (such as blosc, apple accelerate, gcc's libgomp, opencv...) will cause the children process to have a copy of the pthread pool instance in an inconsistent state: the pool metadata will believe that `n > 1` threads are available in the pool while in fact only the main thread state is copied into the children process memory space when doing POSIX fork.

Therefore it is advisable to use `pthread_atfork` to register a handler to always shutdown any pthread pool before executing a fork so that children process will create their own clean pthread pool when needed later (instead of waiting indefinitely for results of threads that do not exist). This way blosc would be usable safely from any Python program without the runtime overhead induced by `blosc.set_releasegil(True)`.

**Comment by FrancescAlted** [comment]
That's interesting. As always, PR are welcome and all that jazz.  Thanks for pointing this out!
